controller/object_store_controller.go

```diff
@@ -0,0 +1,870 @@
+		return nil
+	}
+
+	if err := osc.checkService(svc); errors.Is(err, ErrObjectStoreServiceNotReady) {
+		return nil
+	}
+
+	if err := osc.checkIngress(ingress); errors.Is(err, ErrObjectStoreIngressNotReady) {
+		return nil
+	}
+
+	localEndpoint := *osc.getLocalEndpointURL(store)
+	store.Status.Endpoints = append(store.Status.Endpoints, localEndpoint.DomainName)
+	store, err = osc.setObjectStoreState(store, longhorn.ObjectStoreStateRunning)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// This function does a short sanity check on the various resources that are
+// needed to operate the object stores. If any of them is found to be
+// unhealthy, the controller will transition the object store to "Error"
+// state, otherwise do nothing.
+func (osc *ObjectStoreController) handleRunning(store *longhorn.ObjectStore) (err error) {
+	if store.Spec.TargetState == longhorn.ObjectStoreStateStopped {
+		osc.setObjectStoreState(store, longhorn.ObjectStoreStateStopping)
+		return nil
+	}
+
+	dpl, err := osc.ds.GetDeployment(store.Name)
+	if err != nil || dpl.Status.UnavailableReplicas > 0 {
+		osc.setObjectStoreState(store, longhorn.ObjectStoreStateError)
+		return err
+	}
+
+	_, err = osc.ds.GetService(osc.namespace, store.Name)
+	if err != nil {
+		osc.setObjectStoreState(store, longhorn.ObjectStoreStateError)
+		return err
+	}
+
+	_, err = osc.ds.GetIngress(osc.namespace, store.Name)
+	if err != nil {
+		osc.setObjectStoreState(store, longhorn.ObjectStoreStateError)
+		return err
+	}
+
+	pvc, err := osc.ds.GetPersistentVolumeClaim(osc.namespace, genPVCName(store))
+	if err != nil || pvc.Status.Phase != corev1.ClaimBound {
+		osc.setObjectStoreState(store, longhorn.ObjectStoreStateError)
+		return err
+	}
+
+	_, err = osc.ds.GetVolume(genPVName(store))
+	if err != nil {
+		osc.setObjectStoreState(store, longhorn.ObjectStoreStateError)
+		return err
+	}
+
+	_, err = osc.ds.GetPersistentVolume(genPVName(store))
+	if err != nil {
+		osc.setObjectStoreState(store, longhorn.ObjectStoreStateError)
+		return err
+	}
+	return nil
+}
+
+func (osc *ObjectStoreController) handleStopping(store *longhorn.ObjectStore) (err error) {
+	dpl, err := osc.ds.GetDeployment(store.Name)
+	if err != nil {
+	} else if *dpl.Spec.Replicas != 0 {
+		dpl.Spec.Replicas = int32Ptr(0)
+		_, err = osc.ds.UpdateDeployment(dpl)
+		return err
+	} else if dpl.Status.AvailableReplicas > 0 {
+		return nil // wait for shutdown
+	}
+
+	osc.setObjectStoreState(store, longhorn.ObjectStoreStateStopped)
+	return nil
+}
+
+func (osc *ObjectStoreController) handleStopped(store *longhorn.ObjectStore) (err error) {
+	if store.Spec.TargetState == longhorn.ObjectStoreStateRunning {
+		osc.setObjectStoreState(store, longhorn.ObjectStoreStateStarting)
+		return nil
+	}
+	return nil
+}
+
+func (osc *ObjectStoreController) handleTerminating(store *longhorn.ObjectStore) (err error) {
+	// Just remove the finalizer. The K8s API will take care of removing dependent
+	// resources.
```

Comment:
Suggesting to wait for all dependency resources (volume, PV, PVC, Service, Deployment, Secret) to be fully removed before removing the finalizer of ObjectStore CR.